### ISSUE TITLE
2.8.0 release raise ConnectTimeout on not known url

### ISSUE THREAD

**Issue by aarefiev22** [kennethreitz/requests#2811]
2.7.0 and 2.8.0 raises diff exception on not known url:

import requests
requests.get('http://fake')

2.8.0:
requests.exceptions.ConnectTimeout: HTTPConnectionPool(host='fake', port=80): Max retries exceeded with url: / (Caused by NewConnectionError('<requests.packages.urllib3.connection.HTTPConnection object at 0x7fac9c476110>: Failed to establish a new connection: [Errno -2] Name or service not known',))

2.7.0
requests.exceptions.ConnectionError: ('Connection aborted.', gaierror(-2, 'Name or service not known'))

**Comment by Lukasa** [comment]
Hmm, looks like urllib3 changed how it wrapped this exception. Let me see if I can find where this problem got introduced.

**Comment by Lukasa** [comment]
I suspect shazow/urllib3#703 may have introduced a new exception type that urllib3 treats differently to `socket.error`.

**Comment by sigmavirus24** [comment]
@Lukasa if the OpenStack gate is indicative of anything, it's between 1.10.4 and 1.11

**Comment by Lukasa** [comment]
Looks to me like it was actually done in 1.12: `NewConnectionError` is a subclass of `ConnectTimeoutError`, whereas previously `socket.error` wouldn't have been.

**Comment by aarefiev22** [comment]
@Lukasa : yes looks like new NewConnectionError raised here: https://github.com/kennethreitz/requests/blob/master/requests/packages/urllib3/connection.py#L142

**Comment by Lukasa** [comment]
Hmm, that's not right though: `socket.error` should have entered the retry block.

**Comment by Lukasa** [comment]
Regardless, the reason we think it's a `ConnectTimeout` is because the new exception is a child of `ConnectTimeoutError`. This is somewhat tricky to deal with: `NewConnectionError` is _sometimes_ a timeout, but sometimes not.

**Comment by Lukasa** [comment]
I think from requests' perspective we just want to treat `NewConnectionError` as _not_ a timeout, even though under the hood urllib3 will attempt to retry when it hits it.

**Comment by sigmavirus24** [comment]
Also cc'ing @shazow

**Comment by Lukasa** [comment]
Ok, a proposed bugfix is in #2812. There's some discussion in there about whether this fixes a permanent bug, or whether it's strictly a compatibility fix.

**Comment by shazow** [comment]
I suspect some of this might be platform/configuration-specific. I can't remember the exact details but I think there was a distinction between how OSX vs Linux (maybe Posix vs GNU?) systems behave when you make a new socket to a bad host. It might also have to do with the DNS settings. I vaguely remember in one case it would try to resolve it and hang around until timeout, in the other case it would quickly eliminate it as not resolvable and return a lower-level error.

I tried to normalize them under the same exception but not sure if I succeeded.
